KEP-127: kubelet: honor kubelet user mappings

giuseppe opened the pull request:
allow to specify what IDs must be used by the kubelet to create user namespaces.

If no additional UIDs/GIDs are not allocated to the "kubelet" user, then the kubelet assumes it can use any ID on the system.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature


#### What this PR does / why we need it:

implement a feature defined in the KEP 127 to move to Beta

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
It is possible to configure the IDs that the Kubelet uses to create user namespaces.


User namespaces support is a Beta feature now.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->

```docs
KEP: 127
```

giuseppe committed:
KEP-127: kubelet: honor kubelet user mappings

allow to specify what IDs must be used by the kubelet to create user
namespaces.

If no additional UIDs/GIDs are not allocated to the "kubelet" user,
then the kubelet assumes it can use any ID on the system.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

rata committed:
kublet/userns: Add test switching feature gate off/on

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata commented on pkg/kubelet/userns/userns_manager.go:
```diff
@@ -130,16 +135,35 @@ func (m *UsernsManager) readMappingsFromFile(pod types.UID) ([]byte, error) {
 }
 
 func MakeUserNsManager(kl userNsPodsManager) (*UsernsManager, error) {
+	kubeletMappingID, kubeletMappingLen, err := kl.GetKubeletMappings()
+	if err != nil {
+		return nil, err
+	}
+
+	// If there are no explicit mappings configured, then reserve the first block for the host.
+	off := 1
+	len := math.MaxUint32 / userNsLength
+
+	if kubeletMappingLen > 0 {
```
If we make `kl.GetKubeletMappings()` return the "default" mapping when no custom one is found? It seems simpler. Let's say it returns the range 65536 to MaxID or something like that.

And here we just treat it the same way, no matter if it was configured by the getsubids or none was found, for us here this just returns the range to use.

giuseppe replied:
good idea.  I've changed it

rata commented on pkg/kubelet/userns/userns_manager.go:
```diff
@@ -212,7 +236,7 @@ func (m *UsernsManager) allocateOne(pod types.UID) (firstID uint32, length uint3
 	klog.V(5).InfoS("new pod user namespace allocation", "podUID", pod)
 
-	firstID = uint32(firstZero * userNsLength)
+	firstID = uint32((firstZero + m.off) * userNsLength)
```
Doesn't the bitmap that is already created with offset and len account for the offset? Isn't already taken into account in what it returns, and we are adding it again here?

giuseppe replied:
the bitmap records starting at off=0

rata commented on pkg/kubelet/userns/userns_manager.go:
```diff
@@ -229,7 +253,10 @@ func (m *UsernsManager) record(pod types.UID, from, length uint32) (err error) {
 		return fmt.Errorf("different user namespace range already used by pod %q", pod)
 	}
-	index := int(from / userNsLength)
+	index := int(from/userNsLength) - m.off
```
idem

rata replied:
This is fixed

rata commented on pkg/kubelet/userns/userns_manager.go:
```diff
@@ -229,7 +253,10 @@ func (m *UsernsManager) record(pod types.UID, from, length uint32) (err error) {
 	}
-	index := int(from / userNsLength)
+	index := int(from/userNsLength) - m.off
+	if index >= m.len {
```
Grrr, this shouldn't happen if the API for the bitmap was better. We shouldn't care about this, it should just fail to allocate one if there is no more room or something :-/

rata replied:
Same brain-fart as before, no issue here

rata reviewed (commented):
@giuseppe awsome! I left some simple comments, I hope there is a better way to use the allocator API but I don't know. I'll review in more detail later, here is the first batch :)

rata commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -76,8 +80,71 @@ const (
+func (kl *Kubelet) getKubeletMappings() (uint32, uint32, error) {
+	// default mappings to return if there is no specific configuration
+	const defaultFirstID = 1 << 16
+	const defaultLen = 1<<32 - defaultFirstID
```
The len isn't `1<< 16` (i.e. 65536) too? That seems to be the last valid UID, not the len of each range.

It seems to be used as the length of each userns later too

rata replied:
Grr, I was tired, it is the count since that ID. So this is indeed correct, sorry for the noise again.

giuseppe committed:
features: promote UserNamespacesSupport to beta

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

rata commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -76,8 +80,71 @@ const (
+}
+
+// Get the additional IDs allocated for the Kubelet.
+func (kl *Kubelet) getKubeletMappings() (uint32, uint32, error) {
```
We need to return the default mapping if the feature gate is disabled. Otherwise this can happen:

1. Wrong configuration of the kubelet user is done (i.e. not multiple of userns length)
2. The feature gate is disabled
3. Kubelet initialization still fails

giuseppe replied:
fixed

rata committed:
pkg/kubelet/userns: Remove outdated test

When we were alocating the whole UID space, the first range was reserved
to the host. Now we don't allocate the whole UID space, but just the
range configured, so the first range doesn't point to [0;65535] anymore,
so no need to test it is always set.

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata committed:
kubelet/userns: Remove alpha maxPods limitation

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata committed:
kubelet/userns: Use kubelet maxPods

We don't have the alpha limitation anymore, let's just use the kubelet
maxPods instead of our hardcoded 1024 max.

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata committed:
kubelet/userns: Add unit test

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata commented on pkg/kubelet/userns/userns_manager.go:
```diff
@@ -130,16 +135,35 @@ func (m *UsernsManager) readMappingsFromFile(pod types.UID) ([]byte, error) {
+
+	// If there are no explicit mappings configured, then reserve the first block for the host.
+	off := 1
+	len := math.MaxUint32 / userNsLength
```
Solved now

rata commented on pkg/kubelet/userns/userns_manager_test.go:
```diff
@@ -34,6 +34,11 @@ import (
 
+const (
+	minimumMappingUID = (1 << 16)
+	mappingLen        = (1 << 16) * 2000
```
Fixed

rata commented on pkg/kubelet/userns/userns_manager_test.go:
```diff
@@ -97,6 +106,8 @@ func TestUserNsManagerAllocate(t *testing.T) {
 		assert.Equal(t, userNsLength, int(length), "length is not the expected. iter: %v", i)
 		assert.NoError(t, err)
+		assert.True(t, allocated >= minimumMappingUID)
+		assert.True(t, allocated < minimumMappingUID+mappingLen-userNsLength)
```
Randomly allocated range made it fail, it needs to be <=. Fixed now

mrunalp commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -76,8 +80,75 @@ const (
+	kubeletUser = "kubelet"
 )
 
+func parseGetSubIdsOutput(input string) (uint32, uint32, error) {
```
Can we add some unit tests for this function?

mrunalp replied:
Also, can add a comment above to show example content we are trying to parse.

giuseppe replied:
done

mrunalp commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -76,8 +80,75 @@ const (
+// Get the additional IDs allocated for the Kubelet.
+func (kl *Kubelet) getKubeletMappings() (uint32, uint32, error) {
+	// default mappings to return if there is no specific configuration
+	const defaultFirstID = 1 << 16
```
Can we add some comments here as well?

giuseppe replied:
sure, done

mrunalp commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -76,8 +80,75 @@ const (
+	}
+	cmd, err := exec.LookPath("getsubids")
+	if err != nil {
+		if os.IsNotExist(err) {
```
Can we add a warning here that we can't find the binary and falling back to default mappings?

giuseppe replied:
done

giuseppe commented:
the pull-kubernetes-node-e2e-containerd failure is related to the missing version `gcr.io/cadvisor/cadvisor:v0.47.2`

@bobbypage was it removed?

mrunalp commented:
/test pull-kubernetes-node-e2e-containerd

giuseppe commented:
all green!

haircommander commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -76,8 +80,90 @@ const (
+	if string(outUids) != string(outGids) {
+		return 0, 0, fmt.Errorf("mismatched subuids and subgids for user %q", kubeletUser)
+	}
+	return parseGetSubIdsOutput(string(outUids))
```
nit: if you need to do a repush for anything can the functions be called above where they're defined (switch these two)?

haircommander commented on pkg/kubelet/userns/userns_manager.go:
```diff
@@ -52,14 +47,19 @@ type userNsPodsManager interface {
+	usedBy  map[types.UID]uint32 // Map pod.UID to range used
+	removed int
+
+	off int
+	len int
```
nit: these variable names are a little terse for my taste

haircommander commented:
a couple of nits,  all of which could be follow ups
/lgtm

haircommander commented:
/triage accepted
/priority important-soon
